geopandas/vectorized.pyx

```diff
@@ -1544,29 +1490,41 @@ cpdef cysjoin(np.ndarray[np.uintp_t, ndim=1, cast=True] left,
     cdef size_t left_size = left.size
     cdef size_t right_size = right.size
 
-    handle = get_geos_context_handle()
-    predicate = get_prepared_predicate(predicate_name)
-    if not predicate:
-        raise NotImplementedError(predicate_name)
+    with get_handle() as handle:
+        predicate = get_prepared_predicate(predicate_name)
+        if not predicate:
+            raise NotImplementedError(predicate_name)
 
-    with nogil:
-        sv = sjoin(handle, predicate,
-                   <GEOSGeometry*> left.data, left_size,
-                   <GEOSGeometry*> right.data, right_size)
+        with nogil:
+            sv = sjoin(handle, predicate,
+                       <GEOSGeometry*> left.data, left_size,
+                       <GEOSGeometry*> right.data, right_size)
 
-    left_out = np.empty(sv.n // 2, dtype=np.uintp)
-    right_out = np.empty(sv.n // 2, dtype=np.uintp)
+        left_out = np.empty(sv.n // 2, dtype=np.uintp)
+        right_out = np.empty(sv.n // 2, dtype=np.uintp)
 
-    with nogil:
-        for idx in range(0, sv.n // 2):
-            left_out[idx] = sv.a[2 * idx]
-            right_out[idx] = sv.a[2 * idx + 1]
+        with nogil:
+            for idx in range(0, sv.n // 2):
+                left_out[idx] = sv.a[2 * idx]
+                right_out[idx] = sv.a[2 * idx + 1]
+
+        cfree(sv.a)
 
-    cfree(sv.a)
     return left_out, right_out
 
 
 def concat(L):
     L = list(L)
     x = np.concatenate([ga.data for ga in L])
     return GeometryArray(x, base=set(L))
+
+
+cdef class get_handle:
```

Comment:
I have a slight preference to call this `get_geos_handle`, if it is not too much work to change

Reply:
Excellent @mrocklin ! I confess that I was hand-waving about the context manager and hadn't written any working code :)

Reply:
I had the good fortune of running into a Cython core developer last night, which made this much easier to deal with :)

Reply:
Ah, the benefit of german conferences! :-)